contracts/04-tamagotchi-shop/io/src/lib.rs

```diff
@@ -1,40 +1,236 @@
 #![no_std]
 
-use gmeta::Metadata;
-use gstd::prelude::*;
+use gmeta::{Metadata, In, Out, InOut};
+use gstd::{prelude::*, msg, ActorId, exec};
+
+use sharded_fungible_token_io::{ FTokenAction, FTokenEvent, LogicAction };
+use store_io::{ StoreAction, StoreEvent };
+
+pub type TransactionId = u64;
+pub type AttributeId = u32;
+
+pub const HUNGER_PER_BLOCK: u64 = 1;
+pub const ENERGY_PER_BLOCK: u64 = 2;
+pub const BOREDOM_PER_BLOCK: u64 = 2;
+pub const FILL_PER_SLEEP: u64 = 1000;
+pub const FILL_PER_FEED: u64 = 1000;
+pub const FILL_PER_ENTERTAINMENT: u64 = 1000;
 
 #[derive(Default, Encode, Decode, TypeInfo)]
 #[codec(crate = gstd::codec)]
 #[scale_info(crate = gstd::scale_info)]
 pub struct Tamagotchi {
     // TODO: 0️⃣ Copy fields from previous lesson and push changes to the master branch
+    pub name: String,
+    pub date_of_birth: u64,
+    pub owner: ActorId,
+    pub fed: u64,
+    pub fed_block: u64,
+    pub entertained: u64,
+    pub entertained_block: u64,
+    pub rested: u64,
+    pub rested_block: u64,
+    pub approved_account: Option<ActorId>,    
+        
     // TODO: 2️⃣ Add new fields
+    pub ft_contract_id: ActorId,
+    pub transaction_id: TransactionId,
+    // Check pending transaction
+    pub approve_transaction: Option<(TransactionId, ActorId, u128)>,
+}
+
+impl Tamagotchi {   
+    pub fn sleep(&mut self) {
+        let blocks_height = blocks_height();
+        let updated_rested = updated_field_value(
+            self.rested,
+            self.rested_block,
+            ENERGY_PER_BLOCK,
+            blocks_height
+        );
+        self.rested = update_field(updated_rested, FILL_PER_SLEEP);
+        self.rested_block = blocks_height;  
+    }
+    
+    pub fn feed(&mut self) {
+        let blocks_height = blocks_height();
+        let updated_feed = updated_field_value(
+            self.fed,
+            self.fed_block,
+            HUNGER_PER_BLOCK,
+            blocks_height
+        );
+        self.fed = update_field(updated_feed, FILL_PER_FEED);
+        self.fed_block = blocks_height;
+    }
+    
+    pub fn play(&mut self) {
+        let blocks_height = blocks_height();
+        let updated_entertainer = updated_field_value(
+            self.entertained,
+            self.entertained_block,
+            BOREDOM_PER_BLOCK,
+            blocks_height
+        );
+        self.entertained = update_field(updated_entertainer, FILL_PER_ENTERTAINMENT);
+        self.entertained_block = blocks_height;  
+    }
+    
+    pub fn is_owner_or_approved(&self, user: &ActorId) -> bool {
+        if self.owner == *user {
+            return true;
+        }
+        if self.approved_account == Some(*user) {
+            return true;
+        }
+        false
+    }
+    
+    pub async fn buy_attribute(&mut self, store_id: ActorId, attribute_id: AttributeId) {
+        let response = msg::send_for_reply_as::<_, StoreEvent>(
```

Comment:
You may avoid using turbo-fish here.
```suggestion
        let response: StoreEvent = msg::send_for_reply_as(
```